clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -136,105 +136,344 @@ bool BoundsAnalysis::IsDeclOperand(const Expr *E) {
+  // for all ntptrs in scope should be the same.
+
+  // For example:
+  // _Nt_array_ptr<T> p : bounds(p, p + i);
+  // _Nt_array_ptr<T> q : bounds(p, p + i);
+  // _Nt_array_ptr<T> r : bounds(r, r + i);
+
+  // if (*(p + i)) // widen p and q by 1
+  //   if (*(p + i + 1)) // widen p and q by 2
+  //     if (*(p + i + 2)) // widen p and q by 3
+
+  // if (*(p + i)) // widen p and q by 1
+  //   if (*(p + i + 2)) // no widening
+
+  // if (*p) // no widening
+  //   if (*(p + 1)) // no widening
+  //     if (*(p + i)) // widen p and q by 1
+
+  // TODO: Currently, Lexicographic::CompareExpr does not understand
+  // commutativity of operations. Exprs like "p + e" and "e + p" are considered
+  // unequal.
+
+  for (const VarDecl *V : NtPtrsInScope) {
+    // We try to be conservative in the sense that whenever we find a null
+    // VarDecl or an empty RangeBoundsExpr, we simply skip widening for that
+    // ntptr instead of asserting.
+    if (!V)
+      continue;
+
+    // In case the bounds expr for V is not a RangeBoundsExpr, invoke
+    // ExpandBoundsToRange to expand it to RangeBoundsExpr.
+    const BoundsExpr *BE = S.ExpandBoundsToRange(V, V->getBoundsExpr());
+    const auto *RBE = dyn_cast<RangeBoundsExpr>(BE);
+
+    if (!RBE)
+      continue;
+
+    // Collect all variables involved in the upper and lower bounds exprs for
+    // the ntptr. An assignment to any such variable would kill the widenend
+    // bounds for the ntptr.
+    if (!SuccEB->BoundsVars.count(V)) {
+      DeclSetTy BoundsVars;
+      CollectBoundsVars(RBE, BoundsVars);
+      SuccEB->BoundsVars[V] = BoundsVars;
     }
+
+    // Update the bounds of p on the edge EB->SuccEB only if we haven't already
+    // updated them.
+    if (EB->Gen[SuccEB->Block].count(V))
+      continue;
+
+    const Expr *UE = RBE->getUpperExpr();
+    // Can this ever happen?
+    if (!UE)
+      continue;
+
+    // Split the upper bounds expr into base and offset for matching with the
+    // DerefBase.
+    ExprPairTy UpperExprPair = SplitIntoBaseOffset(UE);
+    const Expr *UpperBase = UpperExprPair.first;
+    const Expr *UpperOffset = UpperExprPair.second;
+    if (!UpperBase)
+      continue;
+
+    if (Lex.CompareExpr(DerefBase, UpperBase) !=
+        Lexicographic::Result::Equal)
+      continue;
+
+    const llvm::APInt DerefOffsetVal = GetAPIntVal(DerefOffset);
+    const llvm::APInt UpperOffsetVal = GetAPIntVal(UpperOffset);
+
+    // We cannot widen the bounds if the offset in the deref expr is less than
+    // the offset in the upper bounds expr. For example:
+    // _Nt_array_ptr<T> p : bounds(p, p + i + 2); // Upper Bounds Offset (UBO) = 2.
+
+    // if (*(p + i)) // Offset = 0 < UBO ==> no widening
+    //   if (*(p + i + 1)) // Offset = 1 < UBO ==> no widening
+    //     if (*(p + i + 2)) // Offset = 2 == UBO ==> widen by 1
+    //       if (*(p + i + 3)) // Offset = 3 > UBO ==> widen by 2
+    if (Lex.CompareAPInt(DerefOffsetVal, UpperOffsetVal) ==
```

Comment:
I am handling this while computing the union. We only widen p if the bounds of p in (In - Kill) == Gen[p]. So we would not widen the bounds by 4 in case of `if (*(p + i + 5)`.

Reply:
I have added a test where widening should not happen because of a missing test of an element.